postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java

```diff
@@ -35,43 +65,113 @@ public PostgresTesterContainer() {
    */
   @Override
   public void start(String database, String username, String password) {
-    if (postgreSQLContainer != null) {
+    if (primary != null) {
       throw new IllegalStateException("already started");
     }
-    postgreSQLContainer = new PostgreSQLContainer<>(dockerImageName)
+
+    boolean configureAsyncCommit = System.getProperty(POSTGRES_ASYNC_COMMIT) != null;
+
+    var replicationSh = Transferable.of(
+        "echo 'synchronous_commit=remote_apply' >> /var/lib/postgresql/data/postgresql.conf\n"
+            + "echo \"synchronous_standby_names='*'\" >> /var/lib/postgresql/data/postgresql.conf\n"
+            + "echo 'host replication replicator 0.0.0.0/0 trust' >> /var/lib/postgresql/data/pg_hba.conf\n"
+            + "psql -U \"$POSTGRES_USER\" -d \"$POSTGRES_DB\" -c 'CREATE USER replicator WITH REPLICATION'");
+
+    if (configureAsyncCommit) {
+      replicationSh = Transferable.of(
+          "echo 'host replication replicator 0.0.0.0/0 trust' >> /var/lib/postgresql/data/pg_hba.conf\n"
+              + "psql -U \"$POSTGRES_USER\" -d \"$POSTGRES_DB\" -c 'CREATE USER replicator WITH REPLICATION'");
+    }
```

Comment:
I can't agree. I have a hard time seeing the value of this PR if it only provides synchronous replication. Why not just keep a single container then? Just throwing an error if you try to write to the reader is somewhat useful, but being able to test  async is more useful IMO.

My opinion is that _any_ FOLIO code that assumes that a write is immediately available in a read replica should be revisited. Having this option gives us the ability to test these necessary changes. The solution is simple, if you need the data right away, use the read/write endpoint not the reader one.

Asynchronous replication is the default for postgres replication and requires extra configuration to make it synchronous. Also if hosters use AWS RDS, which many do, read replica replication is _always_ asynchronous. This results in unexpected replication behavior in production.

From: https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_ReadRepl.html

![image](https://github.com/folio-org/raml-module-builder/assets/104455/5c87665a-1e37-462e-a2f1-332db8ece76b)

Reply:
I won't disagree with the points you're making but I don't see why it should stop us from having an *option* to test an async configuration in RMB. Lets say we're only going to support eventual consistency in new APIs, as you suggest. Wouldn't it be nice to have a way to test those APIs? That's all that is at stake here IMO. I think whether we have to revisit old code is out of scope for this.

Reply:
> The default for replicas must be synchronous because all existing code relies on consistency.

The default _is_ synchronous here, not async. Not tests will fail and people looking at it can be referred to the javadoc. I don't see a risk to providing the async option as an experimental feature.
 
> Therefore async mode can only been used on a per transaction basis.

Async mode _could_ only be done on a transaction basis, except that for anyone using RDS with read replicas and who wants to take advantage of the split, it is configured on the db level, not on the transaction level, and is something they can't change.

A lot of time has been invested in this read write split, and implementers who are using RDS want to take advantage of read replicas, and they should at least have the option of exploring what would be needed to make that work. Without this they don't have that option. Why shut down the _possibility_ of RDS users eventually of taking advantage of the split?

I agree given your assessment of the current code that deploying read replicas in RDS should be done with caution. But the read write split is already merged. We should now provide an option for better testing of this configuration.

Reply:
> Most module code that uses PostgresClient has been written before RMB added the reader option.
> 
> It's in the contract of the PostgresClient methods that written values are immediately available, without eventual consistency

@julianladisch, do the R/W split changes that were implemented on the existing RMB readonly methods continue to uphold this warranty even if the underlying database technology like RDS does not replicate synchronously? 

> There are use cases where performance is important, these can be implemented at new (!) APIs that clearly document that there is only eventual consistency. Code that uses these new APIs is aware of eventual consistency and can take measures to avoid working with inconsistent data.

Here I don't think it matters if the client uses the new APIs or the old APIs because RMB does not have control over how the DB replicates the data. I agree with Steve that there is a need for clients to test their code with the async replication option when R/W split mode is enabled so that they would know all the effects that enabling R/W split could bring.  If unit testing with RMB does not inform clients that there might be an issue due to replication, when they deployed FOLIO into the real world where replication is asynchronous, not until then will an issue be found and it will be too late.  Moreover, if a team has to troubleshoot this kind of an issue (due to async replication) and they cannot reproduce the problem with unit tests, are we not making it more difficult for people to consume this code?  

I agree with you that folks should know what the capabilities of RMB are, what is guaranteed or not, and these things should be documented clearly in the README and elsewhere as appropriate.  It seems to me that not having this option for teams to test (with async replication) will do them a disservice.

Reply:
> Having an async replica in the postgres-testing containers is not sufficient to show that the code under test actually supports eventual consistency resulting from async replication.

Thanks, @julianladisch . I was not suggesting that RMB or our client code must prove that they support eventual consistency. I think Steve concurs with me that we'd want to expose a potential problem due to async replication if one exists. 

> In addition updates must be hold back so that the replica has stale data, this enforced stale data is lacking from the current implementation of PG_ASYNC_COMMIT option. The time lag of the async replica can increase under load in production, therefore it is not sufficient to test with the small default time lag.

So basically you're saying that having this configureAsyncCommit configuration option is useless even to simulate async replication?

Reply:
> Having an async replica in the postgres-testing containers is not sufficient to show that the code under test actually supports eventual consistency resulting from async replication. In addition updates must be hold back so that the replica has stale data, this enforced stale data is lacking from the current implementation of PG_ASYNC_COMMIT option. The time lag of the async replica can increase under load in production, therefore it is not sufficient to test with the small default time lag.

I think this is a valid point Julian, so I looked into how we can introduce a simulated replication delay in the config. There is `recovery_min_apply_delay` which does what we want. I have added it to the async config, introduced a new constant called `SIMULATED_ASYNC_REPLICATION_LAG_MILLISECONDS` and modified the async version of the unit test to make use of it like this:
```
await()
    .atLeast(PostgresTesterContainer.SIMULATED_ASYNC_REPLICATION_LAG_MILLISECONDS - 100, TimeUnit.MILLISECONDS)
    .atMost(PostgresTesterContainer.SIMULATED_ASYNC_REPLICATION_LAG_MILLISECONDS + 100, TimeUnit.MILLISECONDS)
    .until(...);
```
Does this address your concern?